jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java

```diff
@@ -571,61 +574,67 @@ static Chunk next(Chunk chunk)
 
         /**
          * <p>Returns a new {@code Chunk} whose {@code ByteBuffer} is a slice of the
-         * {@code ByteBuffer} of the source {@code Chunk} unless the source
-         * {@link #hasRemaining() has no remaining byte} in which case:</p>
+         * whole {@code ByteBuffer} of the source {@code Chunk} unless the source
+         * {@link #hasRemaining() has no remaining bytes} in which case:</p>
          * <ul>
-         * <li>{@code this} is returned if it is an instance of {@link Error}</li>
-         * <li>{@link #EOF} is returned if {@link #isLast()} is {@code true}</li>
+         * <li>{@code this} is returned if {@link #isLast()} is {@code true}</li>
          * <li>{@link #EMPTY} is returned if {@link #isLast()} is {@code false}</li>
          * </ul>
-         * <p>If the source has remaining bytes, the returned {@code Chunk} retains
-         * the source {@code Chunk} and it is linked to it via
+         * <p>If the source {@code Chunk} has remaining bytes, the returned
+         * {@code Chunk} it is linked to the source {@code Chunk} via
          * {@link #from(ByteBuffer, boolean, Retainable)}.</p>
          *
-         * @return a new {@code Chunk} retained from the source {@code Chunk} with a slice
-         * of the source {@code Chunk}'s {@code ByteBuffer}
+         * @return a new {@code Chunk} with a slice of the source {@code Chunk}'s
+         * {@code ByteBuffer}
          */
         default Chunk slice()
         {
-            if (isTerminal())
-                return this;
-            if (!hasRemaining())
-                return EMPTY;
-            retain();
-            return from(getByteBuffer().slice(), isLast(), this);
+            if (hasRemaining())
+                return from(getByteBuffer().slice(), isLast(), this);
```

Comment:
Shouldn't a retain be done somewhere here? We now have 2 chunks pointing at the same buffer, so 2 releases are now required.   I could not see any retain call in the from method or any of the constructors?

Reply:
No, the retain operation should be (possibly) done by the receiver of the sliced chunk, because it can just consume the sliced chunk synchronously, so there would be no need to retain.
If the consumer of the sliced chunk stores it away for later use, then the consumer is responsible to retain it.

Reply:
We don't allow a caller of read to avoid calling release of they consume the chunk synchronously.

I think it is just simpler to say that if you obtain a new chunk then it must be released.

What if the caller releases the original chunk before finishing with the synchronous processing of the new chunk?

Feels fragile.  If there are performance reasons to avoid the retain/release then we probably need a specially named method with very explicit java doc

Reply:
The change has more a semantic reason, and has performance reasons too.
The places where it is used all consume the slice synchronously, so it felt better to not retain.

The caller should not release the original chunk before consuming the slice, and typically does not because it took a slice of it, meaning that there is something left in the original chunk that needs further processing.

The probability of API usage errors are the same as:
```java
Chunk chunk = read();
chunk.release();
chunk.getByteBuffer().get(0);
```
Yes the API can be used in a way that the chunk is released before being used, but I think it's implicit you should not do that. Same for `slice()`, it's documented that it depends on the original chunk.

Reply:
The sequence I'm concerned about is more like:
```java
    Chunk chunk = read();
    // parse some data
    chunk.position(newPos);
    chunk.limit(newLimit);
    Chunk slice = chunk.slice();
    
    // try to parse more data, but you find none, so ...
    chunk.release();

    // now try to do something with the slice
    slice.get(0); // BOOM! the buffer has already been released because the slice did not retain
```

If slices are consumed synchronously, then why slice at all?
If you really need to do this then have a `sliceWithoutRetain` method, else it is really surprising we allow two chunks to refer to the same buffer, but with a retain counter of only 1!

Reply:
Yours is a wrong sequence, like this one:

```java
Chunk chunk = read();
chunk.release();
chunk.getByteBuffer().get(0); // BOOM!
```

You pointed out in a previous review, when I changed `Chunk.from()` to implicitly retain, that it was wrong because of the silly retain/release cycle, and I agreed it was silly, and it was making `Chunk.from()` a "magic" method because it was, like `read()`, a method that implicitly retains and there was no reason to have 2 "magic" methods.

Here is similar, there is no point to do a silly retain/release cycle and have `slice()` to be a "magic" method.
In this way, there only is one "magic" method: `read()`.

For example, multipart would have to do this in your proposal:

```java
// Implicit retain.
Chunk partContent = chunk.slice(pos, lim, last);

// Give the part content to the application.
onPartContent(partContent);

// Silly release for an implicit retain() 2 lines above.
// The application may have fully consumed 
// the partContent by saving it to disk.
// Plus a performance killer for no reason.
partContent.release();
```

We don't know what the application is doing with the slice, and it's responsibility of the application to retain it if it does not consume it -- not a responsibility of `slice()`.

We document that the slice shares the lifecycle of the original chunk, like we would have to document that it does not and you have to release it if your idea is followed.
So the 2 solutions are equivalent wrt to document their semantic/usage and both rely on documentation and knowing how to use a chunk and therefore write correct code.

Reply:
You want to rename `slice()` to `view()` and keep the current semantic?

We do not need the view+retain semantic, so there won't be a "slice" with retain semantic, because it won't be used.

Reply:
That will help. The normal slice semantic creates an independent reference to the same buffer. Actions on one cannot affect the other (unless the buffer if written to).

With this current semantic, the method is creating a dependent view to the same buffer. Releasing the original may invalidate that view. So it's not a slice semantic.

If a slice (with retain) is not needed, then ok not to have the method.

Reply:
Sorry, I'm unresolving.  It's OK for the from method to not have a retain, but I think the `slice` definitely needs the retain, as we get 2 chunks pointing to the same buffer, but with a reference count of 1.

Reply:
Note that the only time this method is called, it does the retain anyway!

Reply:
Sorry for the noise.  I will resolve this and make a new comment with less verbosity.